cardano-api/src/Cardano/Api/IPC/Monad.hs

```diff
@@ -17,54 +22,138 @@ import Control.Monad.Trans.Cont
+  tmvResultChainTip <- newEmptyTMVarIO
+
+  let waitResult = (,)
+        <$> readTMVar tmvResultChainTip
+        <*> readTMVar tmvResultLocalState
+
+  connectToLocalNodeWithVersion
+    connectInfo
+    (\ntcVersion ->
+      LocalNodeClientProtocols
+      { localChainSyncClient    = LocalChainSyncClient $ chainSyncGetCurrentTip waitResult tmvResultChainTip
+      , localStateQueryClient   = Just $ setupLocalStateQueryExpr waitResult mpoint tmvResultLocalState (f ntcVersion)
+      , localTxSubmissionClient = Nothing
+      }
+    )
 
-          , Net.Query.recvMsgFailure = \failure -> do
-              atomically $ putTMVar resultVar' (Just (Left failure))
+  atomically waitResult
+
+  where
+    chainSyncGetCurrentTip
+      :: STM b
+      -> TMVar  ChainTip
+      -> ChainSyncClient (BlockInMode mode) ChainPoint ChainTip IO ()
+    chainSyncGetCurrentTip waitDone tipVar =
+      ChainSyncClient $ pure clientStIdle
+      where
+        clientStIdle :: Net.Sync.ClientStIdle (BlockInMode mode) ChainPoint ChainTip IO ()
+        clientStIdle =
+          Net.Sync.SendMsgRequestNext clientStNext (pure clientStNext)
+
+        clientStNext :: Net.Sync.ClientStNext (BlockInMode mode) ChainPoint ChainTip IO ()
+        clientStNext = Net.Sync.ClientStNext
+          { Net.Sync.recvMsgRollForward = \_block tip -> ChainSyncClient $ do
+              void . atomically $ putTMVar tipVar tip
+              void $ atomically waitDone
+              pure $ Net.Sync.SendMsgDone ()
+          , Net.Sync.recvMsgRollBackward = \_point tip -> ChainSyncClient $ do
+              void . atomically $ putTMVar tipVar tip
               void $ atomically waitDone
-              pure $ Net.Query.SendMsgDone ()
+              pure $ Net.Sync.SendMsgDone ()
           }
-      else do
-        atomically $ putTMVar resultVar' Nothing
+
+-- | Use 'queryExpr' in a do block to construct monadic local state queries.
+setupLocalStateQueryExpr ::
+     STM x
+  -> Maybe ChainPoint
+  -> TMVar (Either Net.Query.AcquireFailure a)
+  -> LocalStateQueryExpr (BlockInMode mode) ChainPoint (QueryInMode mode) () IO a
+  -> Net.Query.LocalStateQueryClient (BlockInMode mode) ChainPoint (QueryInMode mode) IO ()
+setupLocalStateQueryExpr waitDone mPointVar' resultVar' f =
+  LocalStateQueryClient . pure . Net.Query.SendMsgAcquire mPointVar' $
+    Net.Query.ClientStAcquiring
+    { Net.Query.recvMsgAcquired = runContT (runLocalStateQueryExpr f) $ \result -> do
+        atomically $ putTMVar resultVar' (Right result)
+        void $ atomically waitDone
+        pure $ Net.Query.SendMsgRelease $ pure $ Net.Query.SendMsgDone ()
+
+    , Net.Query.recvMsgFailure = \failure -> do
+        atomically $ putTMVar resultVar' (Left failure)
         void $ atomically waitDone
         pure $ Net.Query.SendMsgDone ()
+    }
+
+-- | Use 'queryExpr' in a do block to construct monadic local state queries.
+queryExpr :: Monad m => query a -> LocalStateQueryExpr block point query r m a
```

Comment:
Let's also concretize `query a` to `QueryInMode mode result` and m to `IO` in order to make the usage of `queryExpr` a bit more obvious.

Reply:
Done!